src/engine/Join.cpp

```diff
@@ -491,3 +492,212 @@ void Join::appendCrossProduct(const IdTable::const_iterator& leftBegin,
+  LOG(DEBUG) << "B: width = " << b.cols() << ", size = " << b.size() << "\n";
+
+  // Check trivial case.
+  if (a.size() == 0 || b.size() == 0) {
+    return;
+  }
+
+  IdTableStatic<OUT_WIDTH> result = dynRes->moveToStatic<OUT_WIDTH>();
+  // Cannot just switch l1 and l2 around because the order of
+  // items in the result tuples is important.
+  if (a.size() / b.size() > GALLOP_THRESHOLD) {
+    doGallopInnerJoin(LeftLargerTag{}, a, jc1, b, jc2, &result);
+  } else if (b.size() / a.size() > GALLOP_THRESHOLD) {
+    doGallopInnerJoin(RightLargerTag{}, a, jc1, b, jc2, &result);
+  } else {
+    // Intersect both lists.
+    size_t i = 0;
+    size_t j = 0;
+    // while (a(i, jc1) < sent1) {
+    while (i < a.size() && j < b.size()) {
+      while (a(i, jc1) < b(j, jc2)) {
+        ++i;
+        if (i >= a.size()) {
+          goto finish;
+        }
+      }
+
+      while (b(j, jc2) < a(i, jc1)) {
+        ++j;
+        if (j >= b.size()) {
+          goto finish;
+        }
+      }
+
+      while (a(i, jc1) == b(j, jc2)) {
+        // In case of match, create cross-product
+        // Always fix a and go through b.
+        size_t keepJ = j;
+        while (a(i, jc1) == b(j, jc2)) {
+          result.push_back();
+          const size_t backIndex = result.size() - 1;
+          for (size_t h = 0; h < a.cols(); h++) {
+            result(backIndex, h) = a(i, h);
+          }
+
+          // Copy bs columns before the join column
+          for (size_t h = 0; h < jc2; h++) {
+            result(backIndex, h + a.cols()) = b(j, h);
+          }
+
+          // Copy bs columns after the join column
+          for (size_t h = jc2 + 1; h < b.cols(); h++) {
+            result(backIndex, h + a.cols() - 1) = b(j, h);
+          }
+
+          ++j;
+          if (j >= b.size()) {
+            // The next i might still match
+            break;
+          }
+        }
+        ++i;
+        if (i >= a.size()) {
+          goto finish;
+        }
+        // If the next i is still the same, reset j.
+        if (a(i, jc1) == b(keepJ, jc2)) {
+          j = keepJ;
+        }
+      }
+    }
+  }
+finish:
+  *dynRes = result.moveToDynamic();
+
+  LOG(DEBUG) << "Join done.\n";
+  LOG(DEBUG) << "Result: width = " << dynRes->cols()
+             << ", size = " << dynRes->size() << "\n";
+}
+
+// _____________________________________________________________________________
+template <typename TagType, int L_WIDTH, int R_WIDTH, int OUT_WIDTH>
+void Join::doGallopInnerJoin(const TagType, const IdTableStatic<L_WIDTH>& l1,
+                             const size_t jc1, const IdTableStatic<R_WIDTH>& l2,
+                             const size_t jc2,
+                             IdTableStatic<OUT_WIDTH>* result) {
+  LOG(DEBUG) << "Galloping case.\n";
+  size_t i = 0;
+  size_t j = 0;
+  while (i < l1.size() && j < l2.size()) {
+    if constexpr (std::is_same<TagType, RightLargerTag>::value) {
+      while (l1(i, jc1) < l2(j, jc2)) {
```

Comment:
I think you missed removing this loop which still does a linear scan. Also it doesn't appear in the `LeftLargerTag` case

Reply:
That does the linear scan in the smaller list. In the `LeftLargerTag` case the loop appears at the end (as it iterates l2 instead of l1)

Reply:
Ahh I see. That makes sense.